pkg/bootflow/examples/bruteforce/main.go

```diff
@@ -0,0 +1,131 @@
+package main
+
+import (
+	"bytes"
+	"crypto/sha1"
+	"fmt"
+	"hash"
+
+	"github.com/9elements/converged-security-suite/v2/pkg/bootflow/bootengine"
+	"github.com/9elements/converged-security-suite/v2/pkg/bootflow/steps/tpmsteps"
+	"github.com/9elements/converged-security-suite/v2/pkg/bootflow/trustchains/tpm"
+	"github.com/9elements/converged-security-suite/v2/pkg/bootflow/types"
+	"github.com/9elements/converged-security-suite/v2/pkg/bruteforcer"
+	"github.com/google/go-tpm/tpm2"
+)
+
+func main() {
+	// Here we brute-force a TPM measurement against a final PCR0 value.
+	//
+	// We initially have measurements of two separators, but we need to
+	// bruteforce the first measurement to 0x01000001
+
+	// an artificial flow for two separators
+	myFlow := types.Flow{
+		tpmsteps.InitTPMLazy(0),
+		tpmsteps.MeasureSeparator(0),
+		tpmsteps.MeasureSeparator(0),
+	}
+
+	// the expected final hash (after bruteforcing)
+	expectedHash := make([]byte, sha1.Size)
+	extend := func(h hash.Hash, a, b []byte) []byte {
+		defer h.Reset()
+		h.Write(a)
+		h.Write(b)
+		return h.Sum(nil)
+	}
+	sep0, sep1 := sha1.Sum([]byte("\x01\x00\x00\x01")), sha1.Sum([]byte("\x00\x00\x00\x00"))
+	expectedHash = extend(sha1.New(), expectedHash, sep0[:])
+	expectedHash = extend(sha1.New(), expectedHash, sep1[:])
+
+	// executing the flow (with two simple "\x00\x00\x00\x00" measurements)
+	state := types.NewState()
+	state.SetFlow(myFlow, 0)
+	state.EnableTrustChain(tpm.NewTPM())
+	process := bootengine.NewBootProcess(state)
+	process.Finish()
+
+	// just some debugging
+	fmt.Printf("Log:\n%#v\n", process.Log)
+
+	// == now let's bruteforce the first measurement ==
+
+	// first find the original (UNHASHED) data of the first measurement
+	tpmMeasurements := process.Log.GetVerifiedDataByTrustChainType((*tpm.TPM)(nil))
+
+	// now let's gather the rest chain (TPM init and the rest TPM Extend-s)
+	var (
+		tpmLocality *uint8
+		tpmExtends  []tpm.CommandLogEntryExtend
+	)
+	tpm.StateExec(state, func(t *tpm.TPM) error {
```

Comment:
this linear search here to extract individual parts from the measure chain seems brittle to me and is different from the design where we had a list of tpm measurements in the sense that, in that case, we could just use the list directly without having knowledge of the flow and alter one item from it in the bruteforcer. In this case, it's basically required to recreate the flow in code (lower down in the bruterforce func) to do the same thing

Reply:
> in that case, we could just use the list directly without having knowledge of the flow and alter one item from it in the bruteforcer.

Here is the same. We do not know the flow, we cherry-pick the first measurement (the same as earlier) and brute force it. The differences are:
1. We manually set TPM locality, but again we extract it from the flow (and we do not need to know the flow to do that).
2. We manually do the "Extend" chain instead of calling `Calculate`. In other words, the internals of function `Calculate` are partly reimplemented in the brute-forcer.

But the brute-forcer is agnostic to the specific flow: the only requirement is to have at least the one measurement, because we brute-force it -- but it was the case in the old design as well.

Reply:
The bruteforcer is agnostic, yes. But this call still implies knowledge about the flow by virtue of having to know that there is a TPM locality initializer in it (or in a non-trivial example, some other list of measurements). It still seems to me that you have a mirror of the flow in the bruteforce callback func, in order to compute the final hash to compare, whereas in the old design, the Calculate method dealt with this impl aspect, rather than the bruteforcer callback impl (and the Calculate encapsulated the flow logic).
Then again, maybe im misinterpreting something.

Reply:
Talked with @mimir-d: we decided to reuse the TPM Extend logic instead of duplicating it.

Reply:
Done. Here is the example: https://github.com/9elements/converged-security-suite/blob/a22e7ec0c568acf6c45271250efa9f9ce2827c0e/pkg/bootflow/examples/bruteforce/main.go